### PR TITLE
feat: Add WhereHas Method to ORM

### PR DESCRIPTION
**Pull request by sCuz12**
Closes # 

## 📑 Description
This PR introduces a feature inspired by Laravel's ORM: the WhereHas method. The WhereHas method allows users to efficiently query records based on the existence of related records and apply conditions to those **related** records.

#### Why it is important ?

- Simplified Complex Queries: With WhereHas, developers can easily perform complex queries involving relationships without the need for multiple database queries or manual filtering.
- Consistency: Aligning with Laravel's ORM conventions makes it easier for developers familiar with Laravel to work with this ORM.

### Example
Suppose you have a database with two main models: User and Book, and there is a one-to-many relationship between users and books, where each user can have multiple books. You want to find users who have books with a specific condition, such as books with the word "fantasy" in their names.

``` 
query.WhereHas("books", func(subQuery ormcontract.Query) ormcontract.Query  {
    // Apply conditions to the related "books" model.
    subQuery.Where("Name", "like", "%fantasy%")
})
``` 

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x ] All the tests have passed

### CODE REVIEW

File: contracts/database/orm/orm.go [hunk 1]
```diff
@@ -96,6 +96,8 @@ type Query interface {
 	OrWhere(query any, args ...any) Query
 	// OrWhereIn adds an "or where column in" clause to the query.
 	OrWhereIn(column string, values []any) Query
+
+	WhereHas(table string, fk string, conditions func(Query) Query ) Query
```
Comment (hwbrzzl) [inline contracts/database/orm/orm.go:100]: S: Please order the method based on the first letter and add a description.

File: database/gorm/query.go [hunk 2]
```diff
@@ -672,6 +672,22 @@ func (r *QueryImpl) OrWhereIn(column string, values []any) ormcontract.Query {
...
+	relatedModelQuery := NewQueryImplByInstance(r.instance.Table(table), r)
+
+	//wrap and modify query with the condition passed from function
+    modifiedQuery := condition(relatedModelQuery)
+
+    // Extract the *gorm.DB instance from the modified QueryImpl
+    subQuery := modifiedQuery.(*QueryImpl).instance.Select(fk)
+
+    // Use the subQuery the main query
+    tx := r.instance.Where("id IN (?)", subQuery)
```
Comment (hwbrzzl) [inline database/gorm/query.go:686]: S: Please format the code.

**Review by hwbrzzl — COMMENTED** [review]
Great work, please solve conflicts.

File: database/gorm/query.go [hunk 3]
```diff
@@ -701,6 +701,22 @@ func (r *QueryImpl) OrWhereIn(column string, values []any) ormcontract.Query {
...
+	relatedModelQuery := NewQueryImplByInstance(r.instance.Table(table), r)
+
+	//wrap and modify query with the condition passed from function
+	modifiedQuery := condition(relatedModelQuery)
+
+	// Extract the *gorm.DB instance from the modified QueryImpl
+	subQuery := modifiedQuery.(*QueryImpl).instance.Select(fk)
+
+	// Use the subQuery the main query
+	tx := r.instance.Where("id IN (?)", subQuery)
```
Comment (hwbrzzl) [inline database/gorm/query.go:715]: S: The foreign key may not be `id`, it will be error in this case.

File: database/gorm/query.go [hunk 4]
```diff
@@ -701,6 +701,22 @@ func (r *QueryImpl) OrWhereIn(column string, values []any) ormcontract.Query {
 	return r.OrWhere(fmt.Sprintf("%s IN ?", column), values)
 }
 
+func (r *QueryImpl) WhereHas(table string, fk string, condition func(ormcontract.Query) ormcontract.Query) ormcontract.Query {
```
Comment (hwbrzzl) [inline database/gorm/query.go:704]: S: We'd like to remove the `fk` field.

**Review by hwbrzzl — CHANGES_REQUESTED** [review]
Thanks for your PR! CI failed, could you please install Mockery and regenerate the interface?

```
go install github.com/vektra/mockery/v2@v2.34.2
mockery --all --keeptree --dir=contracts
```

File: contracts/database/orm/orm.go [hunk 5]
```diff
@@ -135,6 +135,9 @@ type Query interface {
 	Where(query any, args ...any) Query
 	// WhereIn adds a "where column in" clause to the query.
 	WhereIn(column string, values []any) Query
+	// WhereHas applies a subquery condition to filter results based on a related table.
+
```
Comment (hwbrzzl) [inline contracts/database/orm/orm.go:139]:
```suggestion
```

File: database/gorm/query.go [hunk 6]
```diff
@@ -701,6 +701,30 @@ func (r *QueryImpl) OrWhereIn(column string, values []any) ormcontract.Query {
...
+
+	relatedModelQuery := NewQueryImplByInstance(r.instance.Model(table), r)
+
+	//wrap and modify query with the condition passed from function
+	modifiedQuery,_ := condition(relatedModelQuery)
+
+	// Extract the *gorm.DB instance from the modified QueryImpl
+	subQuery := modifiedQuery.(*QueryImpl).instance.Select(fk)
+
+	id := database.GetPrimaryField(table)
```
Comment (hwbrzzl) [inline database/gorm/query.go:714]: There is a situation: The foreign key is not the primary id, User can customize it by tag, for example: 

```
type User struct {  
  gorm.Model  Name
  string  CompanyRefer int  
  Company  Company `gorm:"foreignKey:CompanyRefer"`
}

type Company struct {
  ID   int
  Name string
}
```

```
type User struct {
  gorm.Model  
  Name      string
  CompanyID string
  Company   Company `gorm:"references:Code"`
}

type Company struct {
  ID   int
  Code string
  Name string
}
```

**Review by hwbrzzl — COMMENTED** [review]
CI failed

File: database/gorm/query.go (hunk 4 above)
Comment (sCuz12) [reply database/gorm/query.go]: I've found a solution to dynamically retrieve the foreign key field without needing to explicitly pass it. This approach involves passing both the parent model and the child model to the `WhereHas` method. Here's an example of how to use it:

```go
resultQuery, _ := query.WhereHas(User{}, Book{}, func(query ormcontract.Query) (ormcontract.Query, error) {
    modifiedQuery := query.Where("Name = ?", "Book A")
    return modifiedQuery, nil
})
 
```
With this modification, the foreign key field is dynamically determined based on the relationship between the parent and child models, eliminating the need for the `fk` parameter..

**i WOULD like to hear any feedback and any ideas how to improve further this method withou passing the Parent Model (User{})**

**Comment by hwbrzzl** [comment]
If the PR can be reviewed again, please pin me, thanks!

**Comment by sCuz12** [comment]
Hello @hwbrzzl , The PR is now ready for review. However, it seems that the **last two** tests are failing. I'd like to understand the reason behind these failures and address any issues accordingly. The main message of this failing are _"Added line #xxx was not covered by tests"_
Screenshot 2024-02-15 at 9 38 36 AM

File: contracts/database/orm/orm.go [hunk 7]
```diff
@@ -141,6 +141,8 @@ type Query interface {
 	Where(query any, args ...any) Query
 	// WhereIn adds a "where column in" clause to the query.
 	WhereIn(column string, values []any) Query
+	// WhereHas applies a subquery condition to filter results based on a related table.
+	WhereHas(parentModel any, childModel any,conditions func(Query) (Query,error) ) (Query,error)
```
Comment (hwbrzzl) [inline contracts/database/orm/orm.go:145]: I suggest that the parameters are better be passed like Laravel: 

```
$posts = Post::whereHas('comments', function (Builder $query) {
    $query->where('content', 'like', 'code%');
})->get();
```

Therefore, the `WhereHas` can be:

```
facades.Orm().Query().WhereHas("Comments", func(query ormcontract.Query) (ormcontract.Query, error) {
     modifiedQuery := query.Where("Name = ?", "Book A")
     return modifiedQuery, nil
}).Get(&users)
```

File: database/gorm/query.go [hunk 8]
```diff
@@ -705,6 +705,33 @@ func (r *QueryImpl) OrWhereIn(column string, values []any) ormcontract.Query {
...
+	//Extract Primary key
+	primaryKey := database.GetPrimaryField(parentModel)
+	
+	if primaryKey == "" {
+		// Handle the case where the primary key field is not found
+		return nil, errors.New("primary key field not found")
+	}
+	
+	// Use the subQuery the main query
+	tx := r.instance.Where(primaryKey + " IN (?)", subQuery)
```
Comment (hwbrzzl) [inline database/gorm/query.go:730]: This method doesn't support the `Many To Many` relationships.

File: support/database/database.go [hunk 9]
```diff
@@ -55,3 +72,58 @@ func GetIDByReflect(t reflect.Type, v reflect.Value) any {
...
+            return field.Name
+        }
+    }
+
+    // try to find an association field in the child model
+    for i := 0; i < childType.NumField(); i++ {
+        field := childType.Field(i)
+		fmt.Println(field)
+        if field.Type.Kind() == reflect.Ptr && field.Type.Elem() == parentType {
+            return ToSnakeCase(field.Name) + "_id"
```
Comment (hwbrzzl) [inline support/database/database.go:124]: The `_id` suffix is not safe, the foreign id may not ***_id.

**Review by hwbrzzl — COMMENTED** [review]
Thanks for your hard work, Bro! The biggest problem of the PR is not supporting all relationships, only supports some normal cases. We need to support `HasOne`, `HasMany`, `BelongsTo` and `ManyToMany` at the same time.

**Review by Messhias — APPROVED** [review]

**Comment by hwbrzzl** [comment]
Hi @sCuz12 Are you still interested in working on this, please?

**Comment by rusman-plat-d** [comment]
@sCuz12 any update?

**Comment by hwbrzzl** [comment]
Close this first, please reopen it if have any update.

**Pull request closed without merging**